fix(auth): Attach identity and org membership on verified merge

RyanSkonnord opened the pull request:
Previously, users attaching an AuthIdentity to an existing account via idpmigration.py could do so only if they already were a member of the provider's organization and would hit an OrganizationMember.DoesNotExist exception otherwise. Modify the flow to dispatch correctly to handle_attach_identity, creating new entities of both AuthIdentity and OrganizationMember when needed.

https://getsentry.atlassian.net/browse/ER-861

RyanSkonnord committed:
fix(auth): Debug IdP migration flow when user accepts org invite

Prevent OrganizationMember.DoesNotExist exception.

RyanSkonnord committed:
fix(auth): Attach identity and org membership on verified merge

Previously, users attaching an AuthIdentity to an existing account via
idpmigration.py could do so only if they already were a member of the
provider's organization. Modify the flow to dispatch correctly to
handle_attach_identity, creating new entities of both AuthIdentity and
OrganizationMember when needed.

RyanSkonnord commented on src/sentry/auth/helper.py:
```diff
@@ -466,6 +472,8 @@ def handle_unknown_identity(
                     # assume they've confirmed they want to attach the identity
                     op = "confirm"
+            elif is_account_verified:
+                op = "confirm"
```
This seems like it could be at risk of "an attempt to hijack membership of another organization" as mentioned in the comment above. I invite careful review to make sure that's not the case.

My solution is predicated on:

1. By the time the initial login flow reaches `AccountConfirmLink.store_in_redis`, the user has been authenticated by the org's auth provider, which suffices to allow org membership. (We would proceed to `handle_attach_identity` if the Sentry user were authenticated at that time.)
2. The check on `get_verification_value_from_key` suffices to prove that this is the same user who was authenticated in that way.

Per discussion earlier today, it's possible this is an attacker who has gained access to the real user's inbox. But if the attacker can bypass the SSO check *and* could reset that user's Sentry password, they'd get access regardless.

RyanSkonnord commented on src/sentry/web/frontend/idp_email_verification.py:
```diff
@@ -32,3 +26,27 @@ def handle(self, request: Request, key: str) -> HttpResponse:
+    def _recover_org_slug(verification_value):
+        om = AccountConfirmationView._recover_org_member(verification_value)
+        if om is not None:
+            return om.organization.slug
```
Since we have the `organization_id` anyway, it seems we could skip the `OrganizationMember` query and go directly to the `Organization` query. I'm currently erring on the side of leaving the behavior unchanged in the default case.

JoshFerge replied:
it seems fine to me to simply skip the orgmember query.

RyanSkonnord replied:
👍

JoshFerge reviewed (commented):
did we want to add any tests for this case? or were you wanting to confirm approach first?

RyanSkonnord commented:
> did we want to add any tests for this case? or were you wanting to confirm approach first?

Going to add tests tomorrow, but wanted to get review going on the draft early.

JoshFerge commented on src/sentry/auth/helper.py:
```diff
@@ -239,22 +239,25 @@ def handle_attach_identity(
         """
         Given an already authenticated user, attach or re-attach an identity.
         """
+        user = user or self.user
```
should we replace the comment above? since we are potentially using a user passed in from `_get_user`, the user is not necessarily "authenticated" in the traditional sense (they haven't completed the login flow). Maybe "Given an authenticated user, or a user who has confirmed their email, attach or re-attach an identity"?)

JoshFerge replied:
i'm wondering if it would be possible to clean up this a bit if we initialized the helper with a user, either request.user, `_get_user`, or `None`?

RyanSkonnord replied:
Yeah, passing the user is a code smell. My first (bad) idea was to group the two affected methods into a nested helper class, but yet another layer of OOP would be ugly here.

Are you suggesting that we pull the following code from `handle_unknown_identity` into helper initialization?
```python
        if not self.user.is_authenticated:
            acting_user = self._get_user(identity)
            login_form = AuthenticationForm(
                # stuff
            )
        else:
            acting_user = self.user
            login_form = None
```
Later there are some state changes to `acting_user` that would need to be tweaked, but I think I like it.

JoshFerge reviewed (commented):
I think that we could clean up a bunch of edge cases if `_get_user` were to first look for accounts that have an invite / membership to the organization they're trying to log in to. as of now, duplicate accounts are still going to cause issues in this flow.

JoshFerge commented:
> I think that we could clean up a bunch of edge cases if `_get_user` were to first look for accounts that have an invite / membership to the organization they're trying to log in to. as of now, duplicate accounts are still going to cause issues in this flow.

Also, are there any hueristics to determine, if there are duplicates, which account to select that would provide the best user experience? one most recently active? one with the most settings set / most logins?

JoshFerge commented on src/sentry/web/frontend/idp_email_verification.py:
```diff
@@ -32,3 +26,27 @@ def handle(self, request: Request, key: str) -> HttpResponse:
+        except OrganizationMember.DoesNotExist:
+            return None  # the user has left the org since the record was written
+
+    @staticmethod
```
nit: is it cleaner to make this a `@classmethod` so you can use `cls` below instead of having to reference `AccountConfirmationView`?

RyanSkonnord replied:
Good idea, but obviated if we skip the org member query.

JoshFerge commented on src/sentry/auth/helper.py:
```diff
@@ -295,13 +298,13 @@ def handle_attach_identity(
 
         if member is None:
-            member = self._get_organization_member(auth_identity)
+            member = self._get_organization_member(auth_identity, user)
```
i wrote this code and looking at it, it seems a better name is `_get_or_create_organization_member`

JoshFerge reviewed (commented):
overall makes sense to me. I see what you're saying about refactoring -- the previous flow where we check if the account is verified could all be removed and use `handle_attach_identity` instead if we can correctly decide the acting user prior. will let you determine how much you want to bite off here.  

and I think we should pay attention to dupe accounts at this point, since we know it will come up again eventually

RyanSkonnord committed:
Refactor user and identity into AuthIdentityHandler instance variables

Move the operation of looking up an existing user from the email address
in the identity up front to the creation of the AuthIdentityHandler
object. This prevents the need to pass around the "acting user" among
methods later.

Business logic within handle_unknown_identity around "blanking" the
acting user to None, thereby triggering a new account provisioning
(sometimes), is now represented with a more meaningful boolean variable.

RyanSkonnord commented:
> I think that we could clean up a bunch of edge cases if `_get_user` were to first look for accounts that have an invite / membership to the organization they're trying to log in to. as of now, duplicate accounts are still going to cause issues in this flow.

Just to be super clear, this fix includes but is not limited to cases where the user *doesn't* have an invite, and rather is acquiring a membership by passing SSO creds to the org's login portal. So we want to prioritize an account where we *don't* have to create a new `OrganizationMember` row, but still do create a new `OrganizationMember` row where applicable. (I believe this is what you meant; just being explicit.)

> Also, are there any hueristics to determine, if there are duplicates, which account to select that would provide the best user experience? one most recently active? one with the most settings set / most logins?

Beyond the scope of this fix, but that's good for us to think about.

I'd be eager to unify any such logic with [`getsentry.web.identity.HandleIdentityView.handle`](https://github.com/getsentry/getsentry/blob/master/getsentry/web/identity.py), where there's a multi-step flow just to prioritize `Identity`, `AuthIdentity`, and primary/secondary email addresses. 😩

> overall makes sense to me. I see what you're saying about refactoring -- the previous flow where we check if the account is verified could all be removed and use `handle_attach_identity` instead if we can correctly decide the acting user prior. will let you determine how much you want to bite off here.

I'm leaning toward minimal changes to the `handle_attach_identity` flow at this time, but I'd love to attack it as a separate task. Maybe 🤞 the refactoring mentioned [here](https://github.com/getsentry/sentry/pull/30356#discussion_r761542154) will lead to something nice though.

Thanks a lot @JoshFerge, this is great feedback.

RyanSkonnord committed:
Add coverage in test_idpmigration

RyanSkonnord committed:
Merge branch 'master' into fix-idp-migration-without-org-member

RyanSkonnord commented:
See https://github.com/getsentry/sentry/pull/30463 about refactoring the ugly new `user` parameters out from the private methods. I think it's an overall improvement and hope you agree.

I've added test cases to `test_idpmigration` since those are the only places where the presence of the `OrganizationMember` actually affected the flow. I believe the existing test cases in `test_helper` suffice to regression-test the changes in `helper.py`.

JoshFerge commented:
> See #30463 about refactoring the ugly new `user` parameters out from the private methods. I think it's an overall improvement and hope you agree.
> 
> I've added test cases to `test_idpmigration` since those are the only places where the presence of the `OrganizationMember` actually affected the flow. I believe the existing test cases in `test_helper` suffice to regression-test the changes in `helper.py`.

I think some integration test cases in `tests/sentry/web/frontend/test_auth_organization_login.py` would be helpful to add:

- A logged out user has an existing account  with no membership to the current org, with the same email as the one returned from the IDP and heads through the idpmigration flow
- Same as above, but there is an OrganizationMember invite with the same email as the account + idp identity: what do we expect to happen to the invite?

And I remember you saying you considered these below cases out of scope, but I think it will be important to address them as I believe we'll see these issues pop up soon after we release the feature:
 
- A logged out user has multiple existing accounts, where one has membership to the current org
- A logged out user has multiple existing accounts, and none of them have a membership to the current org

RyanSkonnord committed:
Refactor post_init logic into initialize_user

RyanSkonnord commented:
@JoshFerge More tests in `test_auth_organization_login` sounds good to me.

RyanSkonnord committed:
Need _dispatch_to_confirmation to distinguish request user vs existing user

RyanSkonnord committed:
Debug login op

RyanSkonnord committed:
Fix method calls

RyanSkonnord committed:
Re-simplify AuthIdentityHandler's user initialization

RyanSkonnord committed:
Introduce _logged_in_user property

Cuts down on the number of times we have to refer directly to
`self.request.user` and adds a little abstraction around
`is_authenticated` to make internal code cleaner.

RyanSkonnord committed:
Expand OrganizationAuthLoginNoPasswordTest

RyanSkonnord commented:
> * A logged out user has an existing account  with no membership to the current org, with the same email as the one returned from the IDP and heads through the idpmigration flow

Done: https://github.com/getsentry/sentry/pull/30356/commits/fbc8e5887682dc399cf993e5daf90e8837ce49db

The affected part of the flow wasn't actually being exercised in `test_auth_organization_login` because of the way the mocks were set up. I hacked on the mocks and the test covers more now.

> * Same as above, but there is an OrganizationMember invite with the same email as the account + idp identity: what do we expect to happen to the invite?

I expect the behavior would be the same because we match `OrganizationMember` only on the user ID, not email address. The invite would hang around indefinitely afterward, I suppose. Is that a bug?

> * A logged out user has multiple existing accounts, where one has membership to the current org
> * A logged out user has multiple existing accounts, and none of them have a membership to the current org

Right, I want to say both of these would fall into that `User.objects.*.first()` thing, which I consider out of scope because it predates the IdP migration feature. (And has been a technical liability all along, but more so now.)

If a non-logged-in user had multiple accounts, I don't think organization membership would have ever improved their odds of having the right account picked. Is the use case addressed by this fix likelier to encounter that problem? If so, I'd be up to prioritize some improvements. Let's talk about it tomorrow.

RyanSkonnord committed:
Merge branch 'refactor-user-into-handler' into fix-idp-migration-without-org-member

JoshFerge commented on tests/sentry/web/frontend/test_auth_organization_login.py:
```diff
@@ -979,30 +977,63 @@ def setUp(self):
+
+        _, message = email.call_args
+        context = message["context"]
+        assert context["organization"] == self.organization.name
```
could this test / a separate one also hit `"sentry-auth-organization"` and check that the login succeeds and member / identity are created in the no membership case? (similar to last parts of `test_flow_verify_and_link_without_password_login_success`).

RyanSkonnord replied:
👍 Done

RyanSkonnord committed:
Add sentry-idp-email-verification check to OrganizationAuthLoginNoPasswordTest cases

RyanSkonnord committed:
Don't actually need mock Redis

RyanSkonnord commented on tests/sentry/web/frontend/test_auth_organization_login.py:
```diff
@@ -993,23 +992,13 @@ def test_flow_verify_and_link_without_password_sends_email(self, email, mock_red
-        assert redis_obj["organization_id"] == self.organization.id
-        assert redis_obj["identity_id"]["id"] == self.user.email
-        assert redis_obj["provider"] == "dummy"
-
```
This Redis mock was taking the place of a mocked reference to the `idpmigration.send_one_time_account_confirm_link` method. Now that we're exercising the back half of the flow, I don't think we need either.

We could keep the Redis mock in place, for the sake of these ☝️ more direct assertions on what's being inserted. But I think the `sentry-idp-email-verification` part adequately exercises the most important parts and it's more natural not to have to mock `mock_redis.return_value.get`.

RyanSkonnord committed:
Merge branch 'master' into fix-idp-migration-without-org-member
